lib/Driver/ToolChains.cpp

```diff
@@ -526,6 +401,185 @@ ToolChain::constructInvocation(const CompileJobAction &job,
+    else
+      llvm_unreachable("We were told to perform a standard compile, "
+                       "but no mode option was passed to the driver.");
+  case types::TY_Swift:
+  case types::TY_dSYM:
+  case types::TY_AutolinkFile:
+  case types::TY_Dependencies:
+  case types::TY_SwiftModuleDocFile:
+  case types::TY_ClangModuleFile:
+  case types::TY_SerializedDiagnostics:
+  case types::TY_ObjCHeader:
+  case types::TY_Image:
+  case types::TY_SwiftDeps:
+  case types::TY_ModuleTrace:
+  case types::TY_TBD:
+  case types::TY_OptRecord:
+    llvm_unreachable("Output type can never be primary output.");
+  case types::TY_INVALID:
+    llvm_unreachable("Invalid type ID");
+  }
+}
+
+void ToolChain::JobContext::addFrontendInputAndOutputArguments(
+    ArgStringList &Arguments, std::vector<FilelistInfo> &FilelistInfos) const {
+
+  switch (OI.CompilerMode) {
+  case OutputInfo::Mode::StandardCompile:
+    assert(InputActions.size() == 1 &&
+           "Standard-compile mode takes exactly one input (the primary file)");
+    break;
+  case OutputInfo::Mode::BatchModeCompile:
+  case OutputInfo::Mode::SingleCompile:
+    break;
+  case OutputInfo::Mode::Immediate:
+  case OutputInfo::Mode::REPL:
+    llvm_unreachable("REPL and immediate modes handled elsewhere");
+  }
+
+  const bool UseFileList = shouldUseInputFileList();
+  const bool MayHavePrimaryInputs = OI.mightHaveExplicitPrimaryInputs(Output);
+  const bool UsePrimaryFileList =
+      MayHavePrimaryInputs && shouldUsePrimaryInputFileList();
+  const bool FilterInputsByType = shouldFilterInputsByType();
+  const bool UseSupplementaryOutputFileList =
+      shouldUseSupplementaryOutputFileList();
+
+  if (UseFileList) {
+    Arguments.push_back("-filelist");
+    Arguments.push_back(getAllSourcesPath());
+  }
+  if (UsePrimaryFileList) {
+    Arguments.push_back("-primary-filelist");
+    Arguments.push_back(getTemporaryFilePath("primaryInputs", ""));
+    FilelistInfos.push_back({Arguments.back(), types::TY_Swift,
+                             FilelistInfo::WhichFiles::PrimaryInputs});
+  }
+  if (!UseFileList || !UsePrimaryFileList) {
+    addFrontendInputArguments(MayHavePrimaryInputs, UseFileList,
+                              UsePrimaryFileList, FilterInputsByType,
+                              Arguments);
+  }
+
+  if (UseSupplementaryOutputFileList) {
+    Arguments.push_back("-supplementary-output-file-map");
+    Arguments.push_back(getTemporaryFilePath("supplementaryOutputs", ""));
+    FilelistInfos.push_back({Arguments.back(), types::TY_INVALID,
+                             FilelistInfo::WhichFiles::SupplementaryOutput});
+  } else {
+    addFrontendSupplementaryOutputArguments(Arguments);
+  }
+}
+
+void ToolChain::JobContext::addFrontendInputArguments(
+    const bool mayHavePrimaryInputs, const bool useFileList,
+    const bool usePrimaryFileList, const bool filterByType,
+    ArgStringList &arguments) const {
+  llvm::StringSet<> primaries;
+
+  if (mayHavePrimaryInputs)
```

Comment:
Nitpick: please put braces around multi-*line* if-statements, even if it's only a single *statement* in the body.

Reply:
Fixed. (I should know better:)